File: src-tauri/src/service/names.rs
//! Docker-style memorable session names: `<adjective>-<noun>` ("blue-sirius").
//!
//! Rust twin of `src/lib/names.ts`. Both sides read the SAME word lists —
//! `src/lib/names.json` is compiled in here via `include_str!` — so a name
//! minted by the backend (when `new_session` receives an empty `name`) comes
//! from exactly the vocabulary the dialog's dice button uses.
//!
//! Collision policy mirrors Docker's `GetRandomName`: draw random pairs
//! until one is free; after `MAX_TRIES` misses fall back to the last pair
//! with a counting suffix (`blue-sirius-2`, `-3`, …).

use std::collections::HashSet;
use std::sync::OnceLock;

use rand::{Rng, RngExt};

const WORDS_JSON: &str = include_str!("../../../src/lib/names.json");

pub const SEPARATOR: &str = "-";
/// Random draws before falling back to a numeric suffix.
pub const MAX_TRIES: usize = 24;
/// Shared-prefix length that makes a pair redundant ("lunar luna").
pub const SAME_ROOT_PREFIX: usize = 4;

/// True when the adjective and noun share a root and read as a stutter.
pub fn same_root(adjective: &str, noun: &str) -> bool {
    adjective
        .bytes()
        .zip(noun.bytes())
        .take_while(|(a, b)| a == b)
        .count()
        >= SAME_ROOT_PREFIX
}

/// Lowercase, every run of non-alphanumerics becomes one `-`, no leading or
/// trailing dash: `"Blue Sirius"` → `"blue-sirius"`. Used to put friendly
/// names in the same slug space as generated pairs (the TS side uses
/// `finalizeBranchSlug`, which agrees on every value the generator emits).
pub fn slugify(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars().flat_map(char::to_lowercase) {
        if c.is_ascii_alphanumeric() {
            out.push(c);
        } else if !out.is_empty() && !out.ends_with('-') {
            out.push('-');
        }
    }
    while out.ends_with('-') {
        out.pop();
    }
    out
}

/// Make a name acceptable to `validate::tmux_name`: `.` and `:` (tmux's
/// target separators) become `-`.
pub fn tmux_safe(s: &str) -> String {
    s.replace(['.', ':'], "-")
}

#[derive(serde::Deserialize)]
struct Words {
    adjectives: Vec<String>,
    nouns: Vec<String>,
}

fn words() -> &'static Words {
    static WORDS: OnceLock<Words> = OnceLock::new();
    WORDS.get_or_init(|| {
        serde_json::from_str(WORDS_JSON).expect("src/lib/names.json is valid — checked by tests")
    })
}

pub fn adjectives() -> &'static [String] {
    &words().adjectives
}

pub fn nouns() -> &'static [String] {
    &words().nouns
}

/// Mint a name not present in `existing` (compared case-insensitively;
/// `existing` should hold slugs — worktree names, tmux-name suffixes).
pub fn generate_name<R: Rng + ?Sized>(existing: &HashSet<String>, rng: &mut R) -> String {
    let taken: HashSet<String> = existing.iter().map(|s| s.to_lowercase()).collect();
    let adj = adjectives();
    let noun = nouns();
    let mut last = String::new();
    for _ in 0..MAX_TRIES {
        let a = &adj[rng.random_range(0..adj.len())];
        let n = &noun[rng.random_range(0..noun.len())];
        // A same-root draw still spends a try, so the budget stays MAX_TRIES.
        if same_root(a, n) {
            continue;
        }
        last = format!("{a}{SEPARATOR}{n}");
        if !taken.contains(&last) {
            return last;
        }
    }
    if last.is_empty() {
        last = format!("{}{SEPARATOR}{}", adj[0], noun[0]);
    }
    let mut n = 2u64;
    loop {
        let candidate = format!("{last}{SEPARATOR}{n}");
        if !taken.contains(&candidate) {
            return candidate;
        }
        n += 1;
    }
}

/// `generate_name` with the thread-local OS-seeded rng.
pub fn generate_name_default(existing: &HashSet<String>) -> String {
    generate_name(existing, &mut rand::rng())
}

/// The suffix a tmux name carries after the project prefix:
/// `dev-<owner>-<repo>--fix-login-term` → `Some("fix-login")`. `None` for a
/// bare `dev-<owner>-<repo>` or a name that does not follow the convention.
pub fn tmux_name_suffix(tmux_name: &str, owner: &str, repo: &str) -> Option<String> {
    let prefix = format!("dev-{owner}-{repo}--");
    let rest = tmux_name.strip_prefix(&prefix)?;
    let rest = rest.strip_suffix("-term").unwrap_or(rest);
    if rest.is_empty() {
        None
    } else {
        Some(rest.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::convert::Infallible;

    /// rand 0.10 dropped `StepRng`; a constant generator is all these tests
    /// need (every draw lands on the same pair).
    struct ConstRng(u64);
    impl rand::TryRng for ConstRng {
        type Error = Infallible;
        fn try_next_u32(&mut self) -> Result<u32, Infallible> {
            Ok(self.0 as u32)
        }
        fn try_next_u64(&mut self) -> Result<u64, Infallible> {
            Ok(self.0)
        }
        fn try_fill_bytes(&mut self, dst: &mut [u8]) -> Result<(), Infallible> {
            dst.fill(self.0 as u8);
            Ok(())
        }
    }

    #[test]
    fn lists_match_frontend() {
        // Same numbers as `src/lib/names.test.ts` asserts. Both read one JSON
        // file, so this can only fail if someone edits the list and forgets
        // to update one of the two assertions — which is the intended tripwire.
        assert_eq!(adjectives().len(), 67);
        assert_eq!(nouns().len(), 124);
    }

    #[test]
    fn every_word_is_lowercase_ascii_3_to_8_letters_and_unique() {
        let mut seen = HashSet::new();
        for w in adjectives().iter().chain(nouns().iter()) {
            assert!(
                w.len() >= 3 && w.len() <= 8 && w.chars().all(|c| c.is_ascii_lowercase()),
                "bad word {w:?}"
            );
            assert!(seen.insert(w.clone()), "duplicate word {w:?}");
        }
    }

    #[test]
    fn generates_adjective_noun_from_lists() {
        let mut rng = ConstRng(0);
        let n = generate_name(&HashSet::new(), &mut rng);
        let (a, b) = n.split_once('-').expect("has separator");
        assert!(adjectives().iter().any(|w| w == a));
        assert!(nouns().iter().any(|w| w == b));
    }

    #[test]
    fn skips_taken_names_and_falls_back_to_numeric_suffix() {
        // A constant rng always draws the same pair.
        let mut rng = ConstRng(0);
        let first = generate_name(&HashSet::new(), &mut rng);
        let mut taken: HashSet<String> = [first.to_uppercase()].into_iter().collect();
        let second = generate_name(&taken, &mut ConstRng(0));
        assert_eq!(second, format!("{first}-2"));
        taken.insert(second);
        let third = generate_name(&taken, &mut ConstRng(0));
        assert_eq!(third, format!("{first}-3"));
    }

    #[test]
    fn never_returns_an_existing_name() {
        let mut existing = HashSet::new();
        for _ in 0..500 {
            let n = generate_name_default(&existing);
            assert!(!existing.contains(&n));
            existing.insert(n);
        }
    }

    #[test]
    fn never_pairs_same_root_words() {
        assert!(same_root("lunar", "luna"));
        assert!(same_root("cosmic", "cosmos"));
        assert!(!same_root("coral", "corona"));
        for _ in 0..500 {
            let n = generate_name_default(&HashSet::new());
            let (a, b) = n.split_once('-').unwrap();
            assert!(!same_root(a, b), "{n}");
        }
    }

    #[test]
    fn adjectives_are_sorted_without_fleet() {
        let mut sorted = adjectives().to_vec();
        sorted.sort();
        assert_eq!(sorted, adjectives());
        assert!(!adjectives().iter().any(|a| a == "fleet"));
    }

    #[test]
    fn slugify_and_tmux_safe() {
        assert_eq!(slugify("Blue Sirius"), "blue-sirius");
        assert_eq!(slugify("  Fix: the login bug! "), "fix-the-login-bug");
        assert_eq!(slugify(""), "");
        assert_eq!(tmux_safe("dev-o-r--v1.2:x"), "dev-o-r--v1-2-x");
    }

    #[test]
    fn tmux_name_suffix_strips_prefix_and_term() {
        assert_eq!(
            tmux_name_suffix("dev-o-r--blue-sirius", "o", "r").as_deref(),
            Some("blue-sirius")
        );
        assert_eq!(
            tmux_name_suffix("dev-o-r--blue-sirius-term", "o", "r").as_deref(),
            Some("blue-sirius")
        );
        assert_eq!(tmux_name_suffix("dev-o-r", "o", "r"), None);
        assert_eq!(tmux_name_suffix("other", "o", "r"), None);
    }
}

File: src/lib/names.ts
// Docker-style memorable session names: `<adjective>-<noun>` ("blue-sirius").
//
// The word lists live in `names.json` and are shared verbatim with the Rust
// twin (`src-tauri/src/service/names.rs`, `include_str!`), so a name minted
// on either side comes from the same vocabulary. Nouns are stars,
// constellations and other celestial bodies; adjectives are colours and
// short evocative qualities. Every word is lowercase ASCII, 3–8 letters, so
// the slug is always a legal git branch name AND a legal tmux session name
// without any further cleaning.
//
// Collision policy (mirrors Docker's `GetRandomName`): draw random pairs
// until one is free; after `MAX_TRIES` misses fall back to the last pair
// with a numeric suffix (`blue-sirius-2`, `-3`, …). Docker appends a random
// digit instead; a counting suffix is preferred here because it reads as
// "the second blue sirius" rather than as noise.
import words from './names.json';

export const ADJECTIVES: readonly string[] = words.adjectives;
export const NOUNS: readonly string[] = words.nouns;
export const SEPARATOR = '-';
/** Random draws before falling back to a numeric suffix. */
export const MAX_TRIES = 24;
/** Shared-prefix length that makes a pair redundant ("lunar luna", "cosmic cosmos"). */
export const SAME_ROOT_PREFIX = 4;

/** True when the adjective and noun share a root and read as a stutter. */
export function sameRoot(adjective: string, noun: string): boolean {
  let i = 0;
  while (i < adjective.length && i < noun.length && adjective[i] === noun[i]) i++;
  return i >= SAME_ROOT_PREFIX;
}

/** Uniform random in [0, 1). Injectable so tests are deterministic. */
export type Rng = () => number;

function pick(list: readonly string[], rng: Rng): string {
  const i = Math.floor(rng() * list.length);
  // Guard against an rng that returns exactly 1 (or NaN) — clamp into range.
  return list[Math.min(Math.max(i, 0), list.length - 1)] ?? list[0];
}

/**
 * Mint a name that is not in `existing`. Comparison is case-insensitive and
 * `existing` is expected to hold slugs (worktree names, tmux-name suffixes,
 * slugified friendly names) — see `takenSlugsForProject` in the dialog.
 */
export function generateName(
  existing: ReadonlySet<string> = new Set(),
  rng: Rng = Math.random,
): string {
  const taken = new Set<string>();
  for (const e of existing) taken.add(e.toLowerCase());
  let last = '';
  for (let i = 0; i < MAX_TRIES; i++) {
    const adjective = pick(ADJECTIVES, rng);
    const noun = pick(NOUNS, rng);
    // A same-root draw still spends a try, so the budget stays MAX_TRIES.
    if (sameRoot(adjective, noun)) continue;
    last = `${adjective}${SEPARATOR}${noun}`;
    if (!taken.has(last)) return last;
  }
  if (!last) last = `${ADJECTIVES[0]}${SEPARATOR}${NOUNS[0]}`;
  // Every draw collided (tiny pool, or a hostile rng). Count up from 2 so the
  // result is still readable and still unique.
  for (let n = 2; ; n++) {
    const candidate = `${last}${SEPARATOR}${n}`;
    if (!taken.has(candidate)) return candidate;
  }
}

/** `blue-sirius` → `blue sirius` (what the friendly-name field shows). */
export function nameWords(slug: string): string {
  return slug.split(SEPARATOR).filter(Boolean).join(' ');
}

/** True when `slug` is `<adjective>-<noun>` (optionally `-<n>`) from the lists. */
export function isGeneratedName(slug: string): boolean {
  const parts = slug.toLowerCase().split(SEPARATOR);
  if (parts.length < 2 || parts.length > 3) return false;
  if (parts.length === 3 && !/^\d+$/.test(parts[2])) return false;
  return ADJECTIVES.includes(parts[0]) && NOUNS.includes(parts[1]);
}

/**
 * The suffix a tmux name carries after the project prefix:
 * `dev-<owner>-<repo>--fix-login-term` → `fix-login`. Returns null when the
 * name is not of that shape (a bare `dev-<owner>-<repo>` or a foreign name).
 */
export function tmuxNameSuffix(tmuxName: string, owner: string, repo: string): string | null {
  const prefix = `dev-${owner}-${repo}--`;
  if (!tmuxName.startsWith(prefix)) return null;
  let rest = tmuxName.slice(prefix.length);
  if (rest.endsWith('-term')) rest = rest.slice(0, -'-term'.length);
  return rest || null;
}
